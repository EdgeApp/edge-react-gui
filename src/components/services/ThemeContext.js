// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { womanTheme } from '../../theme/variables/womanTheme'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(womanTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
