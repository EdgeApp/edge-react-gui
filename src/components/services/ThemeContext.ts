import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig'
import { Theme } from '../../types/Theme'

export type { Theme }
export { cacheStyles }

/**
 * Utility for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
