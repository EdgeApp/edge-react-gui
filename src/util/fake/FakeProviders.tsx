import * as React from 'react'
import { Metrics, SafeAreaProvider } from 'react-native-safe-area-context'
import { Provider } from 'react-redux'
import { applyMiddleware, createStore } from 'redux'
import thunk from 'redux-thunk'

import { rootReducer, RootState } from '../../reducers/RootReducer'

type DeepPartial<T> = T extends object
  ? {
      [P in keyof T]?: DeepPartial<T[P]>
    }
  : T

export type FakeState = DeepPartial<RootState>

interface Props {
  children: React.ReactNode
  initialState?: FakeState
}

export function FakeProviders(props: Props) {
  const { children, initialState = {} } = props

  const store = React.useMemo(() => createStore(rootReducer, initialState as any, applyMiddleware(thunk)), [initialState])
  return (
    <SafeAreaProvider initialMetrics={initialMetrics}>
      <Provider store={store}>{children}</Provider>
    </SafeAreaProvider>
  )
}

// These match what our old react-native-safe-area-view library returned:
const initialMetrics: Metrics = {
  frame: { height: 1334, width: 750, x: 0, y: 0 },
  insets: { bottom: 0, left: 0, right: 0, top: 20 }
}
