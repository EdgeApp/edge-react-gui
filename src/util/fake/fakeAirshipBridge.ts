import { AirshipBridge } from 'react-native-airship'

export const fakeAirshipBridge: AirshipBridge<any> = {
  resolve: value => undefined,
  // eslint-disable-next-line node/handle-callback-err
  reject: error => undefined,
  remove: () => undefined,
  on: (name, callback) => () => undefined,
  onResult: callback => undefined
}
