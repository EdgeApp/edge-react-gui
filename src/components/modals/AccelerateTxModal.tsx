import { EdgeCurrencyWallet, EdgeDenomination, EdgeTransaction } from 'edge-core-js'
import React, { PureComponent } from 'react'
import { Text, View } from 'react-native'
import { AirshipBridge } from 'react-native-airship'

import s from '../../locales/strings'
import { Slider } from '../../modules/UI/components/Slider/Slider'
import { getDisplayDenominationFromState, getExchangeDenominationFromState } from '../../selectors/DenominationSelectors'
import { connect } from '../../types/reactRedux'
import { GuiExchangeRates } from '../../types/types'
import { convertTransactionFeeToDisplayFee } from '../../util/utils'
import { cacheStyles, Theme, ThemeProps, withTheme } from '../services/ThemeContext'
import { ModalFooter, ModalMessage, ModalTitle } from '../themed/ModalParts'
import { ThemedModal } from '../themed/ThemedModal'
import { Tile } from '../tiles/Tile'

interface OwnProps {
  acceleratedTx: EdgeTransaction
  bridge: AirshipBridge<EdgeTransaction | null>
  replacedTx: EdgeTransaction
  wallet: EdgeCurrencyWallet
}
interface StateProps {
  exchangeRates: GuiExchangeRates
}
interface DispatchProps {
  getDisplayDenomination: (pluginId: string, currencyCode: string) => EdgeDenomination
  getExchangeDenomination: (pluginId: string, currencyCode: string) => EdgeDenomination
}
type Props = OwnProps & StateProps & ThemeProps & DispatchProps

interface State {
  error?: Error
  status: 'confirming' | 'sending' | 'sent'
  mounted: boolean
}

export class AccelerateTxModalComponent extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props)

    this.state = {
      error: undefined,
      status: 'confirming',
      mounted: true
    }
  }

  signBroadcastAndSave = async () => {
    const { acceleratedTx, wallet } = this.props

    this.setState({ status: 'sending' })

    try {
      // Sign, broadcast, and save the accelerated transaction
      const signedTx = await wallet.signTx(acceleratedTx)
      await wallet.broadcastTx(signedTx)
      await wallet.saveTx(signedTx)

      if (this.state.mounted) {
        this.setState({ status: 'sent' })
      }
      this.closeModal(signedTx)
    } catch (error: any) {
      if (this.state.mounted) {
        this.closeModal(null, error)
      }
    }
  }

  closeModal = (signedTx: EdgeTransaction | null, err?: Error) => {
    this.setState({
      mounted: false
    })
    if (err != null) this.props.bridge.reject(err)
    else this.props.bridge.resolve(signedTx)
  }

  handleCancel = () => {
    this.closeModal(null)
  }

  handleConfirmation = async () => {
    await this.signBroadcastAndSave()
  }

  getTxFeeDisplay = (edgeTransaction: EdgeTransaction): string => {
    const { exchangeRates, wallet, getDisplayDenomination, getExchangeDenomination } = this.props

    const feeDisplayDenomination = getDisplayDenomination(wallet.currencyInfo.pluginId, wallet.currencyInfo.currencyCode)
    const feeDefaultDenomination = getExchangeDenomination(wallet.currencyInfo.pluginId, wallet.currencyInfo.currencyCode)
    const transactionFee = convertTransactionFeeToDisplayFee(wallet, exchangeRates, edgeTransaction, feeDisplayDenomination, feeDefaultDenomination)

    const feeSyntax = `${transactionFee.cryptoSymbol ?? ''} ${transactionFee.cryptoAmount} (${transactionFee.fiatSymbol ?? ''} ${transactionFee.fiatAmount})`

    return feeSyntax
  }

  render() {
    const { acceleratedTx, bridge, replacedTx, theme } = this.props
    const { error, status } = this.state

    const styles = getStyles(theme)

    const oldFee = this.getTxFeeDisplay(replacedTx)
    const newFee = this.getTxFeeDisplay(acceleratedTx)

    const isSending = status === 'sending'

    return (
      <ThemedModal bridge={bridge} onCancel={this.handleCancel}>
        <ModalTitle>{s.strings.transaction_details_accelerate_transaction_header}</ModalTitle>
        <ModalMessage>{s.strings.transaction_details_accelerate_transaction_instructional}</ModalMessage>
        <View style={styles.container}>
          <Tile type="static" title={s.strings.transaction_details_accelerate_transaction_old_fee_title} body={oldFee} />
          {newFee == null ? null : <Tile type="static" title={s.strings.transaction_details_accelerate_transaction_new_fee_title} body={newFee} />}
        </View>
        {error == null ? null : (
          <View style={styles.error}>
            <Text style={styles.errorText} numberOfLines={3}>
              {error.message}
            </Text>
          </View>
        )}
        <View style={styles.container}>
          <Slider
            disabled={isSending || !!error}
            onSlidingComplete={this.handleConfirmation}
            showSpinner={isSending}
            disabledText={s.strings.transaction_details_accelerate_transaction_slider_disabled}
          />
        </View>
        <ModalFooter onPress={this.handleCancel} />
      </ThemedModal>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  loadingContianer: {
    height: theme.rem(20)
  },
  error: {
    marginVertical: theme.rem(0.5)
  },
  errorText: {
    textAlign: 'center',
    color: theme.negativeText
  },
  loading: {
    flex: 1,
    alignSelf: 'center'
  },
  container: {
    maxHeight: theme.rem(20),
    marginVertical: theme.rem(0.5)
  }
}))

export const AccelerateTxModal = connect<StateProps, DispatchProps, OwnProps>(
  (state, ownProps) => ({
    exchangeRates: state.exchangeRates
  }),
  dispatch => ({
    getDisplayDenomination(pluginId: string, currencyCode: string) {
      return dispatch(getDisplayDenominationFromState(pluginId, currencyCode))
    },
    getExchangeDenomination(pluginId: string, currencyCode: string) {
      return dispatch(getExchangeDenominationFromState(pluginId, currencyCode))
    }
  })
)(withTheme(AccelerateTxModalComponent))
