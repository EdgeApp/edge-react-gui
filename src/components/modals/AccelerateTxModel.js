// @flow

import type { EdgeCurrencyWallet, EdgeDenomination, EdgeSpendInfo, EdgeTransaction } from 'edge-core-js'
import React, { PureComponent } from 'react'
import { ActivityIndicator, Text, View } from 'react-native'
import { type AirshipBridge } from 'react-native-airship'

import { playSendSound } from '../../actions/SoundActions.js'
import { TRANSACTION_DETAILS } from '../../constants/SceneKeys.js'
import s from '../../locales/strings.js'
import { Slider } from '../../modules/UI/components/Slider/Slider.js'
import { getDisplayDenominationFromState, getExchangeDenominationFromState } from '../../selectors/DenominationSelectors.js'
import { type TestProps, connect } from '../../types/reactRedux.js'
import { Actions } from '../../types/routerTypes.js'
import { type GuiExchangeRates } from '../../types/types.js'
import { convertTransactionFeeToDisplayFee } from '../../util/utils.js'
import { showError, showToast, showWarning } from '../services/AirshipInstance.js'
import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'
import { ModalCloseArrow, ModalMessage, ModalTitle } from '../themed/ModalParts.js'
import { ThemedModal } from '../themed/ThemedModal.js'
import { Tile } from '../tiles/Tile'

type Status = 'confirming' | 'sending' | 'sent'

type OwnProps = {
  bridge: AirshipBridge<Status>,
  edgeTransaction: EdgeTransaction,
  wallet: EdgeCurrencyWallet
}
type StateProps = {
  exchangeRates: GuiExchangeRates
}
type DispatchProps = {
  getDisplayDenomination: (pluginId: string, currencyCode: string) => EdgeDenomination,
  getExchangeDenomination: (pluginId: string, currencyCode: string) => EdgeDenomination
}
type Props = OwnProps & StateProps & ThemeProps & DispatchProps & TestProps

type State = {
  edgeUnsignedTransaction?: EdgeTransaction,
  error?: Error,
  status: Status,
  mounted: boolean
}

export class AccelerateTxModelComponent extends PureComponent<Props, State> {
  constructor() {
    super()

    this.state = {
      edgeUnsignedTransaction: undefined,
      error: undefined,
      status: 'confirming',
      mounted: true
    }
  }

  componentDidMount() {
    this.makeRbfTransaction()
  }

  makeRbfTransaction = async () => {
    const { edgeTransaction, wallet } = this.props

    const edgeTransactionSpendTargets = edgeTransaction.spendTargets

    if (edgeTransactionSpendTargets && edgeTransactionSpendTargets.length) {
      // Currency code for the new tx is the same as the first spend target
      const { currencyCode } = edgeTransactionSpendTargets[0]
      // Map the EdgeTransaction.spendTargets type to EdgeSpendTargets
      const spendTargets = edgeTransactionSpendTargets.map(spendTarget => ({
        nativeAmount: spendTarget.nativeAmount,
        publicAddress: spendTarget.publicAddress,
        uniqueIdentifier: spendTarget.uniqueIdentifier
      }))
      const {
        // Use the txid from the replaced transaction as the rbfTxid
        txid: rbfTxid,
        // Copy replaced transaction's metadata and swapInfo to the RBF transaction
        metadata,
        swapData
      } = edgeTransaction

      const edgeSpendInfo: EdgeSpendInfo = {
        currencyCode,
        spendTargets,
        rbfTxid,
        metadata,
        swapData
      }

      try {
        const edgeUnsignedTransaction = await wallet.makeSpend(edgeSpendInfo)

        this.setState({
          edgeUnsignedTransaction
        })
      } catch (error) {
        this.setState({
          error
        })
      }
    } else {
      const error = new Error('Missing spend target data.')
      this.setState({
        error
      })
    }
  }

  signBroadcastAndSaveRbf = async () => {
    const { wallet } = this.props
    const { edgeUnsignedTransaction } = this.state

    if (edgeUnsignedTransaction) {
      let edgeSignedTransaction: EdgeTransaction = edgeUnsignedTransaction

      this.setState({ status: 'sending' })

      try {
        // Sign, broadcast, and save the RBF transaction
        edgeSignedTransaction = await wallet.signTx(edgeUnsignedTransaction)
        edgeSignedTransaction = await wallet.broadcastTx(edgeSignedTransaction)
        await wallet.saveTx(edgeSignedTransaction)

        if (this.state.mounted) {
          playSendSound().catch(error => console.log(error)) // Fail quietly

          this.setState({ status: 'sent' })

          showToast(s.strings.transaction_success_message)

          Actions.replace(TRANSACTION_DETAILS, { edgeTransaction: edgeSignedTransaction })
        } else {
          showWarning(s.strings.transaction_success_message)
        }
      } catch (error) {
        console.log(error)

        if (this.state.mounted) {
          this.setState({ status: 'confirming' })
          showError(error)
        }
      }
    } else {
      throw new Error(s.strings.invalid_spend_request)
    }
  }

  closeModal = () => {
    this.setState({
      mounted: false
    })
    this.props.bridge.resolve(this.state.status)
  }

  handleConfirmation = async () => {
    await this.signBroadcastAndSaveRbf()
    this.closeModal()
  }

  getTxFeeDisplay = (edgeTransaction: EdgeTransaction): string => {
    const { exchangeRates, wallet, getDisplayDenomination, getExchangeDenomination } = this.props

    const feeDisplayDenomination = getDisplayDenomination(wallet.currencyInfo.pluginId, wallet.currencyInfo.currencyCode)
    const feeDefaultDenomination = getExchangeDenomination(wallet.currencyInfo.pluginId, wallet.currencyInfo.currencyCode)
    const transactionFee = convertTransactionFeeToDisplayFee(wallet, exchangeRates, edgeTransaction, feeDisplayDenomination, feeDefaultDenomination)

    const feeSyntax = `${transactionFee.cryptoSymbol ?? ''} ${transactionFee.cryptoAmount} (${transactionFee.fiatSymbol ?? ''} ${transactionFee.fiatAmount})`

    return feeSyntax
  }

  render() {
    const { bridge, edgeTransaction, theme } = this.props
    const { error, status, edgeUnsignedTransaction } = this.state

    const styles = getStyles(theme)

    const oldFee = this.getTxFeeDisplay(edgeTransaction)
    const newFee = edgeUnsignedTransaction != null ? this.getTxFeeDisplay(edgeUnsignedTransaction) : ''

    const isSending = status === 'sending'

    return (
      <ThemedModal bridge={bridge} onCancel={this.closeModal}>
        {edgeUnsignedTransaction || error ? (
          <>
            <ModalTitle>{s.strings.transaction_details_accelerate_transaction_header}</ModalTitle>
            <ModalMessage>{s.strings.transaction_details_accelerate_transaction_instructional}</ModalMessage>
            <View style={styles.container}>
              <Tile type="static" title={s.strings.transaction_details_accelerate_transaction_old_fee_title} body={oldFee} />
              {!!newFee && <Tile type="static" title={s.strings.transaction_details_accelerate_transaction_new_fee_title} body={newFee} />}
            </View>
            {error && (
              <View style={styles.error}>
                <Text style={styles.errorText} numberOfLines={3}>
                  {error.message}
                </Text>
              </View>
            )}
            <View style={styles.container}>
              <Slider
                disabled={isSending || !!error}
                onSlidingComplete={this.handleConfirmation}
                showSpinner={isSending}
                disabledText={s.strings.transaction_details_accelerate_transaction_slider_disabled}
              />
            </View>
            <ModalCloseArrow onPress={this.closeModal} ref={this.props.generateTestHook('AccelerateTxModal.Close')} />
          </>
        ) : (
          <View style={styles.loadingContianer}>
            <ActivityIndicator color={theme.primaryText} style={styles.loading} size="large" />
          </View>
        )}
      </ThemedModal>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  loadingContianer: {
    height: theme.rem(20)
  },
  error: {
    marginVertical: theme.rem(0.5)
  },
  errorText: {
    textAlign: 'center',
    color: theme.negativeText
  },
  loading: {
    flex: 1,
    alignSelf: 'center'
  },
  container: {
    maxHeight: theme.rem(20),
    marginVertical: theme.rem(0.5)
  }
}))

export const AccelerateTxModel = connect<StateProps, DispatchProps, OwnProps & TestProps>(
  (state, ownProps) => ({
    exchangeRates: state.exchangeRates
  }),
  dispatch => ({
    getDisplayDenomination(pluginId: string, currencyCode: string) {
      return dispatch(getDisplayDenominationFromState(pluginId, currencyCode))
    },
    getExchangeDenomination(pluginId: string, currencyCode: string) {
      return dispatch(getExchangeDenominationFromState(pluginId, currencyCode))
    }
  })
)(withTheme(AccelerateTxModelComponent))
