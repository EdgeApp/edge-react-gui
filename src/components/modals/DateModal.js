// @flow

import DateTimePicker from '@react-native-community/datetimepicker'
import * as React from 'react'
import { Appearance, Platform, Text, TouchableOpacity } from 'react-native'
import { type AirshipBridge, AirshipModal } from 'react-native-airship'

import s from '../../locales/strings.js'
import { type ThemeProps, withTheme } from '../services/ThemeContext.js'

export type Props = {
  bridge: AirshipBridge<Date>,
  initialValue: Date
}

type State = {
  darkMode: boolean,
  date: Date
}

/**
 * Shows the native iOS date picker inside a modal.
 */
export class DateModalIos extends React.Component<Props & ThemeProps, State> {
  subscription: { remove(): void } | void

  constructor(props: Props & ThemeProps) {
    super(props)
    this.state = {
      darkMode: Appearance.getColorScheme() === 'dark',
      date: props.initialValue
    }
  }

  componentDidMount() {
    this.subscription = Appearance.addChangeListener(({ colorScheme }) => {
      this.setState({ darkMode: colorScheme === 'dark' })
    })
  }

  componentWillUnmount() {
    if (this.subscription != null) this.subscription.remove()
  }

  /**
   * Wrap the data picker component in an Airship modal.
   * This modal doesn't use the normal theme colors,
   * since the native component inside uses its own phone-based colors.
   */
  render(): React.Node {
    const { bridge, theme } = this.props
    const { darkMode, date } = this.state

    const textStyle = {
      color: darkMode ? theme.dateModalTextDark : theme.dateModalTextLight,
      fontSize: theme.rem(1),
      textAlign: 'right',
      paddingHorizontal: theme.rem(1),
      paddingVertical: theme.rem(0.5)
    }

    return (
      <AirshipModal
        bridge={bridge}
        onCancel={this.handleDone}
        borderRadius={0}
        backgroundColor={darkMode ? theme.dateModalBackgroundDark : theme.dateModalBackgroundLight}
      >
        <TouchableOpacity onPress={this.handleDone}>
          <Text style={textStyle}>{s.strings.string_done_cap}</Text>
        </TouchableOpacity>
        <DateTimePicker display="spinner" mode="date" onChange={this.handleChange} value={date} />
      </AirshipModal>
    )
  }

  handleChange = (event: mixed, date?: Date) => {
    this.setState({ date })
  }

  handleDone = () => {
    this.props.bridge.resolve(this.state.date)
  }
}

/**
 * Displays the native Android date picker modal,
 * using the Airship system to manage its lifetime.
 */
export function DateModalAndroid(props: Props) {
  const { bridge, initialValue } = props

  return (
    <DateTimePicker
      mode="date"
      onChange={(event, date?: Date) => {
        bridge.resolve(date != null ? date : initialValue)
        bridge.remove()
      }}
      value={initialValue}
      // ref for android?
    />
  )
}

export const DateModal = Platform.OS === 'android' ? DateModalAndroid : withTheme(DateModalIos)
