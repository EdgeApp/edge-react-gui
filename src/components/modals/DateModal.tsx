import DateTimePicker from '@react-native-community/datetimepicker'
import * as React from 'react'
import { Appearance, Platform, Text, TextStyle, TouchableOpacity } from 'react-native'
import { AirshipBridge, AirshipModal } from 'react-native-airship'

import { lstrings } from '../../locales/strings'
import { ThemeProps, withTheme } from '../services/ThemeContext'

export interface Props {
  bridge: AirshipBridge<Date>
  initialValue: Date
}

interface State {
  darkMode: boolean
  date: Date
}

/**
 * Shows the native iOS date picker inside a modal.
 */
export class DateModalIos extends React.Component<Props & ThemeProps, State> {
  subscription: { remove: () => void } | undefined

  constructor(props: Props & ThemeProps) {
    super(props)
    this.state = {
      darkMode: Appearance.getColorScheme() === 'dark',
      date: props.initialValue
    }
  }

  componentDidMount() {
    this.subscription = Appearance.addChangeListener(({ colorScheme }) => {
      this.setState({ darkMode: colorScheme === 'dark' })
    })
  }

  componentWillUnmount() {
    if (this.subscription != null) this.subscription.remove()
  }

  /**
   * Wrap the data picker component in an Airship modal.
   * This modal doesn't use the normal theme colors,
   * since the native component inside uses its own phone-based colors.
   */
  render() {
    const { bridge, theme } = this.props
    const { darkMode, date } = this.state

    const textStyle: TextStyle = {
      color: darkMode ? theme.dateModalTextDark : theme.dateModalTextLight,
      fontSize: theme.rem(1),
      textAlign: 'right',
      paddingHorizontal: theme.rem(1),
      paddingVertical: theme.rem(0.5)
    }

    return (
      <AirshipModal
        bridge={bridge}
        onCancel={this.handleDone}
        borderRadius={0}
        backgroundColor={darkMode ? theme.dateModalBackgroundDark : theme.dateModalBackgroundLight}
      >
        <TouchableOpacity onPress={this.handleDone}>
          <Text style={textStyle}>{lstrings.string_done_cap}</Text>
        </TouchableOpacity>
        <DateTimePicker display="spinner" mode="date" onChange={this.handleChange} value={date} />
      </AirshipModal>
    )
  }

  handleChange = (event: unknown, date?: Date) => {
    // @ts-expect-error
    this.setState({ date })
  }

  handleDone = () => {
    this.props.bridge.resolve(this.state.date)
  }
}

/**
 * Displays the native Android date picker modal,
 * using the Airship system to manage its lifetime.
 */
export function DateModalAndroid(props: Props) {
  const { bridge, initialValue } = props

  return (
    <DateTimePicker
      mode="date"
      onChange={(event, date?: Date) => {
        bridge.resolve(date != null ? date : initialValue)
        bridge.remove()
      }}
      value={initialValue}
    />
  )
}

export const DateModal = Platform.OS === 'android' ? DateModalAndroid : withTheme(DateModalIos)
