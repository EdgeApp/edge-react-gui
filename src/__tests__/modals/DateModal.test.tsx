import { describe, expect, it } from '@jest/globals'
import { render } from '@testing-library/react-native'
import * as React from 'react'

import { DateModalIos } from '../../components/modals/DateModal'
import { getTheme } from '../../components/services/ThemeContext'
import { fakeAirshipBridge } from '../../util/fake/fakeAirshipBridge'
import { FakeProviders } from '../../util/fake/FakeProviders'

describe('DateModalIos', () => {
  it('should render with loading props', () => {
    const rendered = render(
      <FakeProviders>
        <DateModalIos
          bridge={fakeAirshipBridge}
          // Chancellor on brink of second bailout for banks:
          initialValue={new Date('2009-01-03T18:15:05.000Z')}
          theme={getTheme()}
        />
      </FakeProviders>
    )

    expect(rendered.toJSON()).toMatchSnapshot()
    rendered.unmount()
  })
})
