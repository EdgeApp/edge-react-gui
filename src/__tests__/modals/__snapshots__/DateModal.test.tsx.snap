// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateModalIos should render with loading props 1`] = `
<AirshipModal
  backgroundColor="#FFFFFF"
  borderRadius={0}
  bridge={
    {
      "on": [Function],
      "onResult": [Function],
      "reject": [Function],
      "remove": [Function],
      "resolve": [Function],
    }
  }
  onCancel={[Function]}
>
  <TouchableOpacity
    onPress={[Function]}
  >
    <Text
      style={
        {
          "color": "#0073D9",
          "fontSize": 22,
          "paddingHorizontal": 22,
          "paddingVertical": 11,
          "textAlign": "right",
        }
      }
    >
      Done
    </Text>
  </TouchableOpacity>
  <Picker
    display="spinner"
    mode="date"
    onChange={[Function]}
    value={2009-01-03T18:15:05.000Z}
  />
</AirshipModal>
`;
